docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+# Data Contracts for CoreCLR, and the data contract stream
+
+Data contracts represent the manner in which a tool which is not the runtime can reliably understand and observe the behavior of the runtime. Contracts are defined by their documentation, and the runtime describes what contracts are applicable to understanding that runtime.
+
+## Data Contract Stream
+The physical layout of this stream is not defined in this document, but its practical effects are.
+
+The data contract stream has a set of records of the following forms.
+
+### Global Values
+Global values which can be of types (int8, uint8, int16, uint16, int32, uint32, int64, uint64, pointer, unicode string)
+All global values have a unicode string describing their name, and a value of one of the above types.
+
+### Compatible Contract
+Each compatible contract is described by a unicode string naming the contract, and a uint32 version. It is an ERROR if multiple versions of a contract are specified in the stream.
+
+### Data Structure Layout
+Each data structure layout has a unicode name for the type, followed by a list of fields. These fields can be of primitive types (int8, uint8, int16, uint16, int32, uint32, int64, uint64, pointer) or of another named data structure type. Each field descriptor provides the offset of the field, the name of the field, and the type of the field.
```

Comment:
I think we also need a total size: the reader might not be interested in accessing every single field of some opaque data structure, but it may need to traverse an array of them.  So it needs to know the total size.  Also some contracts might be in terms of semi-opaque OS specific data structures where we don't actually know the complete set of fields.

Reply:
I agree that the total size is needed for array data traversal and it should be included for data structures that are used in arrays. I do not think it should be mandatory on every structure.

Reply:
> I agree that the total size is needed for array data traversal and it should be included for data structures that are used in arrays. I do not think it should be mandatory on every structure.

That is a fair argument and one I can get behind if we find the 2 or 4 extra bytes for some types are marginally beneficial. I would argument we should focus on creating regular data encoding patterns and then optimize if we determine they are needed. Imagine some compressed encoding, like metadata has, that could be employed which would make data for all contracts consistent with special casing.

Reply:
+1 to total size, that sounds like a great idea to carry around for these things.